src/types/createRouteOptions.ts

```diff
@@ -0,0 +1,126 @@
+  loadComponent: (params: Params) => any,
+}
+
+export function isRouteWithParent(options: CreateRouteOptions): options is CreateRouteOptionsWithParent<Route> {
+  return 'parent' in options
+}
+
+export function isRouteWithComponent(value: CreateRouteOptions): value is CreateRouteOptions & WithComponent {
+  return 'component' in value
+}
+
+export function isRouteWithComponents(value: CreateRouteOptions): value is CreateRouteOptions & WithComponents {
+  return 'components' in value
+}
+
+export function isRouteWithComponentCallback(value: CreateRouteOptions): value is CreateRouteOptions & WithComponentCallback {
+  return 'loadComponent' in value
+}
+
+export function isRouteWithoutComponent(value: CreateRouteOptions): value is Omit<CreateRouteOptions, 'component' | 'components'> {
+  return !isRouteWithComponent(value) && !isRouteWithComponents(value) && !isRouteWithComponentCallback(value)
+}
+
+type ParentPath<TParent extends Route | undefined> = TParent extends Route ? TParent['path'] : Path<'', {}>
+type ParentQuery<TParent extends Route | undefined> = TParent extends Route ? TParent['query'] : Query<'', {}>
+type ComponentCallbackParams<TCombinedPath extends Path, TCombinedQuery extends Query> = ExtractParamTypes<MergeParams<TCombinedPath['params'], TCombinedQuery['params']>>
+
+export type CreateRouteOptions<
+  TName extends string | undefined = string,
+  TPath extends string | Path | undefined = string | Path | undefined,
+  TQuery extends string | Query | undefined = string | Query | undefined,
+  THost extends string | Host | undefined = string | Host | undefined,
+  TParent extends Route | undefined = undefined
+> = Partial<WithComponent | WithComponents | WithComponentCallback<ComponentCallbackParams<CombinePath<ParentPath<TParent>, ToPath<TPath>>, CombineQuery<ParentQuery<TParent>, ToQuery<TQuery>>>>> & WithHooks & {
+  /**
+   * Name for route, used to create route keys and in navigation.
+   */
+  name?: TName,
+  /**
+   * Host part of URL.
+   */
+  host?: THost,
+  /**
+   * Path part of URL.
+   */
+  path?: TPath,
+  /**
+   * Query (aka search) part of URL.
+   */
+  query?: TQuery,
+  /**
+   * Represents additional metadata associated with a route, customizable via declaration merging.
+   */
+  meta?: RouteMeta,
+  // will be removed in follow-up PR
+  disabled?: boolean,
+}
+
+export type CreateRouteOptionsWithoutParent<
+  TName extends string | undefined = undefined,
+  TPath extends string | Path | undefined = undefined,
+  TQuery extends string | Query | undefined = undefined
+> = CreateRouteOptions<TName, TPath, TQuery> & {
+  parent?: never,
+}
+
+export type CreateRouteOptionsWithParent<
+  TParent extends Route,
+  TName extends string | undefined = undefined,
+  TPath extends string | Path | undefined = undefined,
+  TQuery extends string | Query | undefined = undefined
+> = CreateRouteOptions<TName, TPath, TQuery, Host<'', {}>, TParent> & {
+  parent: TParent,
+}
```

Comment:
nit: I'd probably move these into the `createRoute` service same as you did for the external route types.

Reply:
great catch. I actually don't need both. Keeping a version of this here and removing the duplicate type in external